src/demo.html

```diff
@@ -0,0 +1,249 @@
+              <option value="display">Display</option>
+              <option value="music">Music</option>
+              <option value="sensors">Sensors</option>
+            </select>
+            <textarea
+              id="program"
+              rows="10"
+              spellcheck="false"
+              autocapitalize="false"
+              autocomplete="false"
+            >
+            </textarea>
+            <button id="run">Run</button>
+          </div>
+        </div>
+      </div>
+    </div>
+    <script src="term.js"></script>
+    <script>
+      // Connect the simulator REPL to a terminal.
+      const simulator = document.querySelector("#simulator").contentWindow;
+      const term = new Terminal({
+        cols: 80,
+        rows: 24,
+        useStyle: true,
+        screenKeys: true,
+        cursorBlink: false,
+      });
+      term.open(document.getElementById("term"));
+      term.removeAllListeners("data");
+      term.on("data", function (data) {
+        simulator.postMessage(
+          {
+            kind: "serial_input",
+            data,
+          },
+          "*"
+        );
+      });
+
+      const readyCallbacks = [];
+      window.addEventListener("message", (e) => {
+        const { data } = e;
+        if (e.source === simulator) {
+          switch (data.kind) {
+            case "serial_output": {
+              term.write(e.data.data);
+              break;
+            }
+            case "ready": {
+              // We create this here to allow embedders to use
+              // relevant styling/widgets.
+              createSensorUI(data.sensors);
+              while (readyCallbacks.length) {
+                readyCallbacks.pop()();
+              }
+              break;
+            }
+            case "sensor_change": {
+              createSensorUI(data.sensors);
+            }
+          }
+        }
+      });
+
+      // Support for running samples via the REPL.
+      // There is no filesystem support for the moment.
+      const sample = document.querySelector("#sample");
+      const program = document.querySelector("#program");
+      async function fetchProgram(name) {
+        const response = await fetch(`examples/${name}.py`);
+        if (response.ok) {
+          const text = await response.text();
+          program.value = text;
+        } else {
+          program.value = `# No matching example found for ${name}`;
+        }
+      }
+      sample.addEventListener("change", async () => {
+        fetchProgram(sample.value);
+      });
+      fetchProgram(sample.value);
+
+      // Ideally this would just be changing main.py via some filesystem
+      // message, then resetting.
+      document.querySelector("#run").addEventListener("click", async () => {
+        simulator.postMessage(
+          {
+            kind: "serial_input",
+            // Ctrl-C to interrupt, Ctrl-D to reboot (straight to REPL as no program)
+            data: `\x03\x04`,
```

Comment:
I've since learnt that the correct way to do this is via raw mode: https://docs.micropython.org/en/latest/reference/repl.html?highlight=serial%20program#raw-mode-and-raw-paste-mode

I'll update to use this approach in a new PR.